# insertional\_mutagenesis\_long-read

## Long read viral integration processing pipeline
### This pipeline is intended to take fastq files from long-read target-enrichment sequencing (TES) and test for integration between a host and viral vector.
#### Job scripts are written for LSF job scheduler and will need adjusting for other systems

## Input needs

### Sample fastq files, split by sample and deduplicated (recommend pbmarkdup for PacBio data)
### Hybrid host-vector fasta, generated by concatenating chosen host genome with vector sequences relevant to the sample, followed by bwa indexing
#### Optionally, vector\_elems.txt file in viral\_genome directory, containing vector sequence contig names - useful for separating these from haplotype/patch regions in host genome, if present
### file "config.txt" - tab-delimited file containing sample name (fastq prefix) and hybrid genome (fasta prefix)
### Directory structure as follows:
#### parent >
##### 	project >
######		functions
######		viral\_genome
######		00.fastq

## Running analysis
```{bash}

parent=/path/to/parent
project=project

## adjust absolute paths in project-specific bsub scripts

sed -i "/s/parent=.*/parent=${parent}/" ${parent}/${project}/functions/.*
sed -i "/s/project=.*/project=${project}/" ${parent}/${project}/functions/.*


### align, cuteSV

bsub < ${parent}/${project}/functions/run.01.align_cuteSV.bsub

### reconstruct IS - hold on previous job

bsub < ${parent}/${project}/functions/run.02.reconstruct_IS.bsub

### get vector reads - hold on run.01 job

bsub < ${parent}/${project}/functions/run.02.get_vector_reads.bsub


```

## Results

#### project >
####	01.alignment: indexed bam alignment files per sample
####	02.cuteSV: cuteSV vcf results
####	02.single\_source\_reads:
#####		{sample}\_chromosome-counts.txt: txt file with counts per host chromosome, for host-only reads
#####		{sample}\_read-chromosomes.txt: txt file with chromosomes associated with reads, for single-contig reads
#####		{sample}\_vector\_reads.txt: txt file with read IDs for vector-only reads
####	03.reconstruct\_IS:
#####		{sample}\_SVs.tsv: host-vector breakpoint location information per read - each breakpoint within a read will have its own line
#####		{sample}\_IS\_reads.tsv: reconstructed reads, one line per read: read	qwidth	IS\_read	IS\_read\_simplified	IS\_read\_mapq
####	03.vector\_only\_reads: read names, read info, and subset bams for vector-only reads

### Proceed to figure generation and other secondary analyses

